### PR TITLE
re-define callback order

### PR DESCRIPTION
**Pull request by wxchan**
current problem:
callback order: user-defined in callback parameter -> reset_learning_rate/print_evaluation/record_evaluation/early_stop
user can't insert a callback between last 4 callbacks

solution:
set order variable:
reset_learning_rate = 10
print_evaluation = 10
record_evaluation = 20
early_stop = 30
user-defined = some int
default: according to index in callback parameter list, = index - len(callbacks) (all < 0)

current callback order:
before iter: user-defined -> reset_learning_rate -> user-defined
after iter: user-defined -> print_evaluation -> user-defined -> record_evaluation -> user-defined -> early_stop -> user-defined

### CODE REVIEW

File: python-package/lightgbm/engine.py [hunk 1]
```diff
@@ -211,7 +215,7 @@ def train(params, train_data, num_boost_round=100,
         booster.add_valid(valid_set, name_valid_set)
     """start training"""
     for i in range(init_iteration, init_iteration + num_boost_round):
-        for cb in callbacks_before_iter:
+        for cb in sorted(callbacks_before_iter, key=lambda x: x.__dict__.get('order', 0)):
```
Comment (guolinke) [inline python-package/lightgbm/engine.py]: sort this outsize of loop?

**Comment by guolinke** [comment]
Why need to sort callbacks? I think these callbacks are independent.

**Comment by wxchan** [comment]
Some printing order may be different. And we don't know what kind of callbacks user may define.

File: python-package/lightgbm/engine.py (hunk 1 above)
Comment (wxchan) [inline python-package/lightgbm/engine.py]: It should be same in python by constructing a sorted list first.

Comment (wxchan) [inline python-package/lightgbm/engine.py]: ok, fixed

File: python-package/lightgbm/callback.py [hunk 2]
```diff
@@ -108,24 +111,30 @@ def reset_learning_rate(learning_rates):
...
         if isinstance(learning_rates, list):
-            if len(learning_rates) != env.end_iteration:
-                raise ValueError("Length of list 'learning_rates' has to equal 'num_boost_round'.")
-            booster.reset_parameter({'learning_rate':learning_rates[iteration]})
+            if len(learning_rates) != env.end_iteration - env.begin_iteration:
+                raise ValueError("Length of list 'learning_rates' has to equal to 'num_boost_round'.")
+            env.model.reset_parameter({'learning_rate':learning_rates[env.iteration]})
         else:
-            booster.reset_parameter({'learning_rate':learning_rates(iteration, env.end_iteration)})
+            argc = len(inspect.getargspec(learning_rates).args)
```
Comment (wxchan) [inline python-package/lightgbm/callback.py]: @guolinke I don't know xgboost designed learning_rates with 2 arguments here. I made it support both f(curr_iter) and f(curr_iter, total_iter). Is it also related to the feval issue you talked about yesterday?

Comment (guolinke) [inline python-package/lightgbm/callback.py]: ok, I think we can according to this to support better fobj and feval. We can do this in another PR.
